lib/util/result.ts

```diff
@@ -1,73 +1,441 @@
+        return Result.err(nullError);
+      }
+
+      if (result === undefined) {
+        return Result.err(undefinedError);
+      }
+
+      return Result.ok(result);
     } catch (error) {
       return Result.err(error);
     }
   }
 
-  private static wrapPromise<T>(promise: Promise<T>): Promise<Result<T>> {
-    return promise.then(
-      (value) => Result.ok(value),
-      (error) => Result.err(error)
-    );
+  /**
+   * Returns a discriminated union for type-safe consumption of the result.
+   * When `fallback` is provided, the error is discarded and value is returned directly.
+   *
+   *   ```ts
+   *
+   *   // DESTRUCTURING
+   *   const { val, err } = Result.ok('foo').unwrap();
+   *   expect(val).toBe('foo');
+   *   expect(err).toBeUndefined();
+   *
+   *   // FALLBACK
+   *   const value = Result.err('bar').unwrap('foo');
+   *   expect(val).toBe('foo');
+   *
+   *   ```
+   */
+  unwrap(): Res<T, E>;
+  unwrap(fallback: NonNullable<T>): NonNullable<T>;
+  unwrap(fallback?: NonNullable<T>): Res<T, E> | NonNullable<T> {
+    if (fallback === undefined) {
+      return this.res;
+    }
+    return this.res.ok ? this.res.val : fallback;
   }
 
-  static wrap<T>(callback: () => T): Result<T>;
-  static wrap<T>(promise: Promise<T>): Promise<Result<T>>;
-  static wrap<T>(
-    input: (() => T) | Promise<T>
-  ): Result<T> | Promise<Result<T>> {
-    return input instanceof Promise
-      ? Result.wrapPromise(input)
-      : Result.wrapCallback(input);
+  /**
+   * Transforms the value if the result is `ok`.
+   * The transform function can be sync or async.
+   *
+   * It can return plain value, but for better control
+   * of the error type, it's recommended to return `Result`.
+   *
+   *   ```ts
+   *
+   *   // SYNC
+   *   const { val, err } = Result.ok('foo')
+   *     .transform((x) => x.length)
+   *     .unwrap();
+   *   expect(val).toBe(3);
+   *
+   *   // ASYNC
+   *   const { val, err } = await Result.wrap(
+   *     http.getJson('https://api.example.com/data.json')
+   *   )
+   *     .transform(({ body }) => body)
+   *     .unwrap();
+   *
+   *   ```
+   */
+  transform<U, EE>(
+    fn: (value: NonNullable<T>) => Result<U, EE>
+  ): Result<U, E | EE>;
+  transform<U, EE>(
+    fn: (value: NonNullable<T>) => Promise<Result<U, EE>>
+  ): AsyncResult<U, E | EE>;
+  transform<U>(
+    fn: (value: NonNullable<T>) => Promise<NonNullable<U>>
+  ): AsyncResult<U, E | Error>;
+  transform<U>(
+    fn: (value: NonNullable<T>) => NonNullable<U>
+  ): Result<U, E | Error>;
+  transform<U, EE>(
+    fn: (
+      value: NonNullable<T>
+    ) =>
+      | NonNullable<U>
+      | Result<U, EE>
+      | Result<U, EE>
+      | Promise<NonNullable<U>>
+  ): Result<U, E | EE> | AsyncResult<U, E | EE> {
+    if (!this.res.ok) {
+      return Result.err(this.res.err);
+    }
+
+    try {
+      const res = fn(this.res.val);
+
+      if (res instanceof Promise) {
+        return new AsyncResult((resolve) => {
```

Comment:
why using constructor instead of wrap function here?

Reply:
Wrap is silent, any type of errors is assumed.

Transform is verbose, it should handle all errors and specify which errors it's able to return.
Uncaught errors are reported via logging.

Generally speaking, it's an attempt to control which error types are possible at particular code fragment, something JS/TS is missing by design.